README.md
openmediavault-virtualhosts
===========================


debian/changelog
openmediavault-virtualhosts (0.5.2) stable; urgency=medium

  * Increase LimitRequestFieldSize to avoid error 400
  * Stopped postrm from removing configuration when just upgrading
  * Add usage of public subdirectory when creating a virtual host

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 27 Jan 2014 17:18:17 +0100

openmediavault-virtualhosts (0.5.1) stable; urgency=low

  * Add locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 22 Dec 2013 11:48:14 -0500

openmediavault-virtualhosts (0.5) stable; urgency=low

  * Initial release

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 04 Oct 2013 18:55:14 +0200


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/Settings.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/form/Panel.js")

Ext.define("OMV.module.admin.service.virtualhosts.Settings", {
    extend : "OMV.workspace.form.Panel",

    rpcService   : "VirtualHosts",
    rpcGetMethod : "getSettings",
    rpcSetMethod : "setSettings",

    getFormItems : function() {
        var me = this;

        return [{
            xtype    : "fieldset",
            title    : "General settings",
            defaults : {
                labelSeparator : ""
            },
            items : [{
                xtype      : "checkbox",
                name       : "enable",
                fieldLabel : _("Enable"),
                checked    : false
            },{
                xtype         : "numberfield",
                name          : "server_name_default_port",
                fieldLabel    : _("Default ServerName port"),
                vtype         : "port",
                minValue      : 0,
                maxValue      : 65535,
                allowDecimals : false,
                allowNegative : false,
                allowBlank    : false,
                value         : 8000,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Default port to use with name-based VirtualHosts.")
                }]
            }]
        }];
    }
});

OMV.WorkspaceManager.registerPanel({
    id        : "settings",
    path      : "/service/virtualhosts",
    text      : _("Settings"),
    position  : 10,
    className : "OMV.module.admin.service.virtualhosts.Settings"
});


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/VirtualHosts.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")

OMV.WorkspaceManager.registerNode({
    id      : "virtualhosts",
    path    : "/service",
    text    : _("VirtualHosts"),
    icon16  : "images/virtualhosts.png",
    iconSvg : "images/virtualhosts.svg"
});


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/window/Alias.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/window/Form.js")

Ext.define("OMV.module.admin.service.virtualhosts.window.Alias", {
    extend : "OMV.workspace.window.Form",

    title           : _("Add alias"),
    hideResetButton : true,
    mode            : "local",

    getFormItems : function() {
        var me = this;

        return [{
            xtype       : "textfield",
            name        : "alias_url",
            fieldLabel  : _("URL"),
            regex       : /^\/\w+/i,
            invalidText : _("The URL must start with a '/'."),
            allowBlank  : false,
            allowNone   : false,
        },{
            xtype       : "sharedfoldercombo",
            fieldLabel  : _("Shared folder"),
            allowBlank  : true,
            allowNone   : true,
            isFormField : false,
            listeners   : {
                scope    : me,
                'select' : function(combo, records, eOpts) {
                    if (records.length === 0)
                        return;

                    var alias_path = me.findField('alias_path');

                    OMV.Rpc.request({
                        scope    : me,
                        callback : function(id, success, response) {
                            if (success)
                                alias_path.setValue(response);
                        },
                        rpcData  : {
                            service : "ShareMgmt",
                            method  : "getPath",
                            params  : {
                                uuid : records[0].get("uuid")
                            }
                        }
                    });
                }
            },
            plugins    : [{
                ptype : "fieldinfo",
                text  : _("Automatically populate the path with the path of the selected shared folder.")
            }]
        },{
            xtype       : "textfield",
            name        : "alias_path",
            fieldLabel  : _("Path"),
            regex       : /^\/[^\0]+/i,
            invalidText : _("The path must start with a '/'."),
            allowBlank  : false,
            allowNone   : false
        }];
    },

    onOkButton : function() {
        var me = this;

        if (!me.isValid())
            return;

        var url  = me.findField("alias_url").value;
        var path = me.findField("alias_path").value;

        me.store.add({
            alias_url  : url,
            alias_path : path
        });

        me.close();
    }
});


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/VirtualHostList.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/module/admin/service/virtualhosts/window/VirtualHost.js")

Ext.define('OMV.module.admin.service.virtualhosts.VirtualHostList', {
    extend   : "OMV.workspace.grid.Panel",
    requires : [
        "OMV.data.Store",
        "OMV.data.Model",
        "OMV.data.proxy.Rpc",
        "OMV.module.admin.service.virtualhosts.window.VirtualHost"
    ],

    hidePagingToolbar : true,

    columns : [{
        header    : _("UUID"),
        hidden    : true,
        dataIndex : "uuid"
    },{
        header    : _("Enabled"),
        sortable  : true,
        dataIndex : "enable",
        align     : "center",
        width     : 80,
        resizable : false,
        renderer  : OMV.util.Format.booleanIconRenderer(
            "switch_on.png", "switch_off.png")
    },{
        header    : _("Document root"),
        flex      : 1,
        sortable  : true,
        dataIndex : "document_root_full_path"
    },{
        header    : _("URL"),
        flex      : 1,
        sortable  : true,
        dataIndex : "urls",
        renderer  : function(urls) {
            var presentation = "";

            for (var i = 0, j = urls.length; i < j; i++) {
                presentation += urls[i].replace(/!domain!/g, window.location.hostname);

                if (i < j - 1)
                    presentation += ', ';
            }

            return presentation;
        }
    }],

    initComponent : function() {
        var me = this;

        Ext.apply(me, {
            store : Ext.create("OMV.data.Store", {
                autoload   : true,
                remoteSort : false,
                model      : OMV.data.Model.createImplicit({
                    idProperty   : 'uuid',
                    totalPoperty : 'total',
                    fields       : [
                        { name : 'uuid' },
                        { name : 'enable' },
                        { name : 'document_root_full_path' },
                        { name : 'urls' }
                    ]
                }),
                proxy : {
                    type    : 'rpc',
                    rpcData : {
                        "service" : "VirtualHosts",
                        "method"  : "getList"
                    }
                }
            })
        });

        me.doReload();

        me.callParent(arguments);
    },

    onAddButton : function() {
        var me = this;

        Ext.create("OMV.module.admin.service.virtualhosts.window.VirtualHost", {
            title        : _("Add virtual host"),
            listeners    : {
                scope  : me,
                submit : function() {
                    me.doReload();
                }
            }
        }).show();
    },

    onEditButton : function() {
        var me = this;
        var record = me.getSelected();

        Ext.create("OMV.module.admin.service.virtualhosts.window.VirtualHost", {
            rpcGetMethod : "get",
            title        : _("Edit virtual host"),
            uuid         : record.get('uuid'),
            listeners    : {
                scope  : me,
                submit : function() {
                    me.doReload();
                }
            }
        }).show();
    },

    doDeletion : function(record) {
        var me = this;

        OMV.Rpc.request({
            scope : me,
            callback : me.onDeletion,
            rpcData : {
                service : "VirtualHosts",
                method : "delete",
                params : {
                    uuid : record.get("uuid")
                }
            }
        });
    }

});

OMV.WorkspaceManager.registerPanel({
    id        : "virtualhosts",
    path      : "/service/virtualhosts",
    text      : _("Virtual hosts"),
    position  : 20,
    className : "OMV.module.admin.service.virtualhosts.VirtualHostList"
});


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/window/AliasGrid.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/module/admin/service/virtualhosts/window/Alias.js")

Ext.define('OMV.module.admin.service.virtualhosts.window.AliasGrid', {
    extend : 'Ext.grid.Panel',
    uses   : 'OMV.module.admin.service.virtualhosts.window.Alias',

    height : 200,
    stripeRows: true,

    columns : [{
        text     : _("Alias"),
        flex     : 1,
        sortable : false,
        dataIndex: 'alias_url'
    },{
        text     : _("Path"),
        flex     : 1,
        sortable : false,
        dataIndex: 'alias_path'
    }],

    initComponent : function() {
        var me = this;

        me.store = Ext.create('Ext.data.ArrayStore', {
            fields : [
                'alias_url',
                'alias_path'
            ]
        });

        me.dockedItems = [];
        me.dockedItems.push(Ext.widget({
            xtype: "toolbar",
            dock: "top",
            items: me.getTopToolbarItems(me)
        }));

        me.callParent(arguments);

        var selModel = me.getSelectionModel();
        selModel.on('selectionchange', function(model, records) {
            var deleteButton = me.queryById(me.getId() + "-delete");

            if (records.length > 0)
                deleteButton.enable();
            else
                deleteButton.disable();
        });
    },

    getTopToolbarItems : function() {
        var me = this;

        return [{
            id      : me.getId() + "-add",
            xtype   : "button",
            text    : _("Add"),
            icon    : "images/add.png",
            iconCls : Ext.baseCSSPrefix + "btn-icon-16x16",
            handler : me.onAddButton,
            scope   : me
        },{
            id       : me.getId() + "-delete",
            xtype    : "button",
            text     : _("Delete"),
            icon     : "images/delete.png",
            iconCls  : Ext.baseCSSPrefix + "btn-icon-16x16",
            handler  : me.onDeleteButton,
            scope    : me,
            disabled : true
        }];
    },

    onAddButton : function() {
        var me = this;

        Ext.create("OMV.module.admin.service.virtualhosts.window.Alias", {
            store : me.store
        }).show();
    },

    onDeleteButton : function() {
        var me = this;
        var records = me.getSelectionModel().getSelection();

        if (records.length > 0)
            me.store.remove(records[0]);
    }
});


var/www/openmediavault/js/omv/module/admin/service/virtualhosts/window/VirtualHost.js
/**
 * Copyright (C) 2013 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/form/plugin/LinkedFields.js")
// require("js/omv/form/field/SharedFolderComboBox.js")
// require("js/omv/module/admin/service/virtualhosts/window/AliasGrid.js")

Ext.define("OMV.module.admin.service.virtualhosts.window.VirtualHost", {
    extend : "OMV.workspace.window.Form",
    uses   : [
        "OMV.workspace.window.plugin.ConfigObject",
        "OMV.form.plugin.LinkedFields",
        "OMV.module.admin.service.virtualhosts.window.AliasGrid",
    ],

    plugins : [{
        ptype : "configobject"
    },{
        ptype        : "linkedfields",
        correlations : [{
            name : [
                "public_directory"
            ],
            conditions : [{
                name  : "use_public_directory",
                value : true
            }],
            properties : [
                "show",
                "!allowBlank",
                "!readOnly"
            ]
        },{
            name : [
                "server_name",
                "server_name_use_default_port"
            ],
            conditions : [{
                name  : "host_type",
                op    : ">",
                value : 0
            }],
            properties : [
                "show",
                "!allowBlank",
                "!readOnly"
            ]
        },{
            name : [
                "port"
            ],
            conditions : [{
                name  : "host_type",
                value : 1
            },{
                name  : "server_name_use_default_port",
                value : true
            }],
            properties : [
                "hide",
                "allowBlank",
                "readOnly"
            ]
        }]
    }],

    rpcService   : "VirtualHosts",
    rpcSetMethod : "set",

    height          : 600,
    hideResetButton : true,
    uuid            : null,

    getFormItems : function() {
        var me = this;

        return [{
            xtype : "fieldset",
            title : _("General"),
            items : [{
                xtype      : "checkbox",
                name       : "enable",
                fieldLabel : _("Enable"),
                checked    : true
            },{
                xtype      : "sharedfoldercombo",
                name       : "document_root",
                fieldLabel : _("Document root"),
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("The location needs to have at least read permissions for the user/group www-data")
                }]
            },{
                xtype      : "checkbox",
                name       : "use_public_directory",
                fieldLabel : _("Use public directory"),
                checked    : false,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Use if you serve your public files in a subfolder like public_html or public. This folder will be the document root. It also avoids the need to create an additional shared folder in such cases.")
                }]
            },{
                xtype      : "combo",
                name       : "public_directory",
                fieldLabel : _("Public directory"),
                queryMode  : "local",
                store      : Ext.create("Ext.data.ArrayStore", {
                    fields : [
                        "value"
                    ],
                    data   : [
                        [ "public_html" ],
                        [ "public" ],
                    ]
                }),
                displayField  : "value",
                valueField    : "value",
                allowBlank    : false,
                editable      : false,
                triggerAction : "all",
                value         : "public_html"
            },{
                xtype      : "combo",
                name       : "host_type",
                fieldLabel : _("Host type"),
                queryMode  : "local",
                store      : Ext.create("Ext.data.ArrayStore", {
                    fields : [
                        "value",
                        "text"
                    ],
                    data   : [
                        [ 0, _("Port") ],
                        [ 1, "ServerName" ],
                        [ 2, _("Both") ],
                    ]
                }),
                displayField  : "text",
                valueField    : "value",
                allowBlank    : false,
                editable      : false,
                triggerAction : "all",
                value         : 0
            },{
                xtype         : "numberfield",
                name          : "port",
                fieldLabel    : _("Port"),
                vtype         : "port",
                minValue      : 0,
                maxValue      : 65535,
                allowDecimals : false,
                allowNegative : false,
                value         : 8080,
                listeners     : {
                    'change' : function(field, newValue, oldValue) {
                        OMV.Rpc.request({
                            scope : me,
                            callback : function(id, success, response) {
                                if (success) {
                                    if (!response) {
                                        field.markInvalid("This port is already used!");
                                    }
                                }
                            },
                            rpcData : {
                                service : "VirtualHosts",
                                method : "validatePort",
                                params : {
                                    uuid : me.uuid,
                                    port : newValue
                                }
                            }
                        });
                    }
                }
            },{
                xtype      : "textfield",
                name       : "server_name",
                fieldLabel : "ServerName",
                allowBlank : true,
                readOnly   : true,
                hidden     : true
            },{
                xtype      : "checkbox",
                name       : "server_name_use_default_port",
                fieldLabel : _("Use default port for ServerName"),
                checked    : true,
                readOnly   : true,
                hidden     : true,
                plugins    : [{
                    ptype : "fieldinfo",
                    text  : _("Untick to host the ServerName with the portnumber specified in the field port.")
                }]
            }]
        },{
            xtype : "fieldset",
            title : "Aliases",
            items : [{
                xtype      : "gridfield",
                name       : "aliases",
                gridClass  : "OMV.module.admin.service.virtualhosts.window.AliasGrid",
            }]
        },{
            xtype    : "fieldset",
            title    : "Options",
            layout   : "column",
            defaults : {
                columnWidth : 0.5,
                layout      : 'form',
                border      : false,
            },
            items  : [{
                defaults : {
                    hideLabel      : true,
                    labelSeparator : ''
                },
                items : [{
                    xtype      : "checkbox",
                    name       : "options_exec_cgi",
                    boxLabel   : "ExecCGI",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "options_follow_symlinks",
                    boxLabel   : "FollowSymlinks",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "options_includes",
                    boxLabel   : "Includes",
                    checked    : false
                }]
            },{
                defaults : {
                    hideLabel      : true,
                    labelSeparator : ''
                },
                items : [{
                    xtype      : "checkbox",
                    name       : "options_indexes",
                    boxLabel   : "Indexes",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "options_multi_views",
                    boxLabel   : "MultiViews",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "options_symlinks_if_owner_match",
                    boxLabel   : "SymlinksIfOwnerMatch",
                    checked    : false
                }]
            }]
        },{
            xtype : "fieldset",
            title : "AllowOverride",
            layout   : "column",
            defaults : {
                columnWidth : 0.5,
                layout      : 'form',
                border      : false,
            },
            items : [{
                defaults : {
                    hideLabel      : true,
                    labelSeparator : ''
                },
                items : [{
                    xtype      : "checkbox",
                    name       : "allow_override_auth_config",
                    boxLabel   : "AuthConfig",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "allow_override_file_info",
                    boxLabel   : "FileInfo",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "allow_override_indexes",
                    boxLabel   : "Indexes",
                    checked    : false
                }]
            },{
                defaults : {
                    hideLabel      : true,
                    labelSeparator : ''
                },
                items : [{
                    xtype      : "checkbox",
                    name       : "allow_override_limit",
                    boxLabel   : "Limit",
                    checked    : false
                },{
                    xtype      : "checkbox",
                    name       : "allow_override_options",
                    boxLabel   : "Options",
                    checked    : false
                }]
            }]
        },{
            xtype : "fieldset",
            title : _("Extra options"),
            items : [{
                xtype      : "textarea",
                name       : "extra_options",
                fieldLabel : _("Extra options"),
                allowBlank : true
            }]
        }];
    }
});
